tools/preview_new_release.sh

```diff
@@ -0,0 +1,157 @@
+}
+
+while [[ $# -gt 0 ]] && [[ "$1" == "--"* ]]; do
+    opt="$1";
+    shift 1
+    case "$opt" in
+        "--user")
+            user="$1";
+            user_set=1;
+            shift 1;
+            ;;
+        "--version")
+            version="$1";
+            version_set=1;
+            shift 1;
+            ;;
+        "--update-repositories")
+            update_repos=1;
+            ;;
+        "--prefix")
+            prefix="$1";
+            prefix_set=1;
+            shift 1;
+            ;;
+        *)
+            usage;
+            exit 1;
+            ;;
+    esac
+done;
+
+if [ "$user_set" = 0 ] || [ "$version_set" = 0 ]; then
+    usage;
+    exit 1;
+fi;
+
+preview_branch=preview-ocamlformat-$version
+
+if [ "$prefix_set" = 0 ]; then
+    prefix=$HOME/$preview_branch;
+fi;
+
+log_dir=$prefix/logs;
+
+projects=(
+#    'facebook/flow' # not interested in upgrading
+    'facebook/infer'
+    'mirage/alcotest'
+    'mirage/decompress'
+    'mirage/digestif'
+    'mirage/index'
+    'mirage/irmin'
+    'mirage/mirage'
+    'mirage/ocaml-cohttp'
+#    'mirage/ocaml-conduit' # not interested in upgrading
+    'mirage/ocaml-git'
+    'ocaml/dune'
+    'ocaml/ocaml-lsp'
+    'ocaml/odoc'
+    'ocamllabs/dune-release'
+    'ocaml-ppx/ppxlib'
+    'ocsigen/js_of_ocaml'
+    'owlbarn/owl'
+    'realworldocaml/mdx'
+);
+
+function get_main_branch() {
+    gitHeadsDir="$(git rev-parse --show-toplevel)/.git/refs/heads";
+
+    if [ -f "$gitHeadsDir/main" ]; then
+        main_branch='main';
+    elif [ -f "$gitHeadsDir/master" ]; then
+        main_branch='master';
+    else
+        main_branch=$(git remote show origin | grep 'HEAD branch' | cut -d' ' -f5);
+    fi;
+}
+
+rm -rf $log_dir &> /dev/null || true;
+mkdir --parents $log_dir;
+
+for project in "${projects[@]}"; do
+    cd $prefix;
+
+    namespace=`echo $project | cut -d "/" -f 1`;
+    dir=`echo $project | cut -d "/" -f 2`;
+    echo "=> Checking $project";
+
+    if [ ! -d "$dir" ] ; then
+        fork="git@github.com:$user/$dir.git";
+        upstream="git@github.com:$namespace/$dir.git";
+        git clone --recurse-submodules $fork $dir;
+        cd $dir;
+        git remote add upstream $upstream;
+        get_main_branch;
+    else
+        cd $dir;
+
+        if [ "$update_repos" = 1 ]; then
+            get_main_branch;
+            git checkout $main_branch --quiet;
+            git fetch upstream --quiet;
+            git rebase upstream/$main_branch --quiet;
+            git push origin $main_branch --quiet;
+        fi;
+    fi;
+
+    dune=dune;
+
+    if git show-ref --verify --quiet "refs/heads/$preview_branch"; then
+        get_main_branch;
+        git checkout $main_branch --quiet;
+        git branch -D $preview_branch --quiet;
+    fi;
+    git checkout -b $preview_branch --quiet;
+
+    if [ "$project" == "ocaml/dune" ]; then
+        make release;
+        dune=_build/default/bin/dune.exe;
+    fi;
+
+    ocamlformat_file=`readlink -f .ocamlformat`;
```

Comment:
Is here really a project doing that ? Otherwise `sed` can handle this, all the seds can be written as:
```
  sed -i --follow-symlinks -e "1iversion = $version" -e '/^version/d' .ocamlformat
```
What's the point of doing the formatting while the `version` line is removed ? It would be better to leave it as a check.

Reply:
Yes `infer` is doing that.
We need to remove the version because this script's purpose it to make a preview of unreleased versions (right now 0.21.0-XX-YYYYYYYYY, I don't want to rely on having a local tag, I'm too clumsy and would get it pushed by mistake), so we need to disable the version check (we cannot pass --no-version-check to `dune build @fmt`).

Reply:
I see. Perhaps we should have an environment variable for this option ? Otherwise, the script can be improved anyway with `-i --follow-symlinks`.

Reply:
done